runtime/common/src/apis.rs

```diff
@@ -356,6 +356,10 @@ macro_rules! impl_runtime_apis_plus_common {
 					slot: u32,
 					parent_header: &<Block as BlockT>::Header
 				) -> bool {
+					if pallet_migrations::Pallet::<Self>::pending_migration() {
```

Comment:
Is this code executed by the runtime when producing the block (I think not, but I'm not sure)?

Reply:
This would require us to read the state of each migration on each block, and this grows over time with no bound (currently).

Reply:
Why does it grow ? I thought we were removing previous migrations from the list ?
(Also this is mostly executed on the client (before producing the block), but I'm not sure if executed also during block production)

Reply:
> Is this code executed by the runtime when producing the block (I think not, but I'm not sure)?

This is code executed by the client to determine if the node can produce the block

> This would require us to read the state of each migration on each block, and this grows over time with no bound (currently).

This is not on-chain computation, it is client-side.

Reply:
If this is not used by the on-chain computation, where does the can_author is called by the runtime to verify the block author, I forgot ?

Reply:
`can_author` is not called by the runtime to verify the block author, it is called in the client to determine if it should produce the block

The `T::AccountLookup::lookup_account()` which will fail pre-migration still does happen in the runtime in the nimbus::pallet_author_inherent https://github.com/PureStake/nimbus/blob/main/pallets/author-inherent/src/lib.rs#L181-L200 FindAuthor implementation which is used to set the block author in that pallet per block and notify staking which collator produced the block (via `EventHandler::note_author`). This will still need to be fixed or the author won't be set in storage

Reply:
What do you mean "This will still need to be fixed" ? What is wrong with it ?

Reply:
> What do you mean "This will still need to be fixed" ? What is wrong with it ?

If that account lookup done in the runtime fails, then we will not be able to set the author in storage nor notify staking who produced the block. It might be working because it does the account lookup after the migration has executed.

Reply:
> Why does it grow ?

You're right, it wouldn't grow significantly since we prune migrations over time.

This seems like a good approach.

How many nodes will actually attempt to produce a block? I'm wondering if we might introduce networking issues if dozens of nodes are trying to gossip large migration blocks at the same time.

Reply:
Oh, I think this has come up before and the answer is that each node will try to import their own block before gossiping it, which would filter out the ineligible ones.